[WIP][Discussion] Add deprecation policy for API, metadata and feature behaviors. (#9012)
kubevirt/kubevirt: docs/deprecation.md

Author: @vasiliy-ul @rmohr @fabiand @jean-edouard @EdDev @alaypatel07 

Hey all! Thanks a lot for taking part in reviewing this.

I'd like to raise some general questions and would appreciate your opinion regarding them.
I'm wondering by how our deprecation policy should be similar to [Kubernetes' policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/). For example, in Kubernetes, they define 4 general rules to deprecations:

```
Rule #1: API elements may only be removed by incrementing the version of the API group.

Rule #2: API objects must be able to round-trip between API versions in a given release without
information loss, with the exception of whole REST resources that do not exist in some versions.

Rule #3: An API version in a given track may not be deprecated in favor of a less stable API
version.

Rule #4a: API lifetime is determined by the API stability level

Rule #4b: The "preferred" API version and the "storage version" for a given group may not 
advance until after a release has been made that supports both the new version and the previous
version
```

Should we follow all of these rules?
Should we stick to their exact timing? On the one hand our release cycle is now much similar to Kubernetes, but on the other hand I'm not 100% sure that they time definitions for deprecations fit us properly as-is.
Should we support "round-tripping" between versions exactly as Kubernetes does? Do we do something similar to that at the moment with any of our APIs?

Another concern that I have is regarding GA/Beta/Alpha tracks and versioning in general. IIUC, and please correct me if I'm wrong here, in Kubevirt the API versions are fairly arbitrary or at least inconsistent to a specific policy. IOW, every API behaves in a different way without correlations, where some APIs rarely change their versions. For example, IIUC, our "core" API group's version wasn't changed for [about 6 years](https://github.com/kubevirt/kubevirt/blame/main/staging/src/kubevirt.io/api/core/v1/doc.go).

So my question is:
Should we post this policy and start behaving ourselves once it merges with hopes that indeed we would start to properly bump API versions when changes are made? or
Should we merge a relatively generic policy, then create issues and start refactoring so that we would have a better API versioning, then refine this policy in the future?

Reviewer: > So my question is: Should we post this policy and start behaving ourselves once it merges with hopes that indeed we would start to properly bump API versions when changes are made? or Should we merge a relatively generic policy, then create issues and start refactoring so that we would have a better API versioning, then refine this policy in the future?

I'd prefer that we align as much as we can with k8s, that's why I've worked on moving the storage version to `v1` so we can drop `v1alpha3` finally and start thinking about `v2` for the core API and CRDs.

> I'd like to raise some general questions and would appreciate your opinion regarding them. I'm wondering by how our deprecation policy should be similar to [Kubernetes' policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/). For example, in Kubernetes, they define 4 general rules to deprecations:
> 
> ```
> Rule #1: API elements may only be removed by incrementing the version of the API group.

As I said in another thread I don't believe this means bumping major versions but just the next version in the chain. IOW dropping an element from v1beta1 would require v1beta2 or v1. This makes sense to me in KubeVirt.

> Rule #2: API objects must be able to round-trip between API versions in a given release without
> information loss, with the exception of whole REST resources that do not exist in some versions.

I'm not sure about the need to include this support in KubeVirt. Wouldn't it required conversion hooks that write any dropped attributes and their values as annotations on the newer objects so if we ever went back we could rebuild them?

> Rule #3: An API version in a given track may not be deprecated in favor of a less stable API
> version.

Makes sense in KubeVirt IMHO.

> Rule #4a: API lifetime is determined by the API stability level

This forces graduation or removal and is a good thing IMHO.

> Rule #4b: The "preferred" API version and the "storage version" for a given group may not 
> advance until after a release has been made that supports both the new version and the previous
> version
> ```

This would cause us to slow down the moving of our storage versions, I'm not sure it's really worth it.

> Should we follow all of these rules? Should we stick to their exact timing? On the one hand our release cycle is now much similar to Kubernetes, but on the other hand I'm not 100% sure that they time definitions for deprecations fit us properly as-is. Should we support "round-tripping" between versions exactly as Kubernetes does? Do we do something similar to that at the moment with any of our APIs?

Yeah the timeframes might not work for us and could be changed.

As I've said above I don't think round tripping is required at the moment for KubeVirt.

> Another concern that I have is regarding GA/Beta/Alpha tracks and versioning in general. IIUC, and please correct me if I'm wrong here, in Kubevirt the API versions are fairly arbitrary or at least inconsistent to a specific policy. IOW, every API behaves in a different way without correlations, where some APIs rarely change their versions. For example, IIUC, our "core" API group's version wasn't changed for [about 6 years](https://github.com/kubevirt/kubevirt/blame/main/staging/src/kubevirt.io/api/core/v1/doc.go).

I'd argue that this entire effort is aimed at avoiding the mess we got into with the core API in the future and would allow us to move on to `v2` at some point.

> So my question is: Should we post this policy and start behaving ourselves once it merges with hopes that indeed we would start to properly bump API versions when changes are made? or Should we merge a relatively generic policy, then create issues and start refactoring so that we would have a better API versioning, then refine this policy in the future?

Yes I think we should align with k8s, but I don't think this policy would cause us to `bump API versions when changes are made`. As you've said above rule 4.a states that `API lifetime is determined by the API stability level` not when new elements are added.

After the above deprecation process, the feature gate should remain as a no-op with a warning popping up.

### Metadata (e.g. labels / annotations)

> Kubevirt adds certain labels / annotations to both objects that is created by us (e.g. VMs) and cluster-wide

Reviewer 2: ```suggestion
Kubevirt adds certain labels / annotations to both objects that are created by us (e.g. VMs) and cluster-wide
```

Author: done

longer relevant / useful.

#### Deprecation
In order to deprecate metadata the following conditions must be met:
* A mail should be sent to kubevirt-dev mailing list (kubevirt-dev@googlegroups.com) to inform about the deprecation.
* In case the metadata is being renamed / changed, the new metadata (e.g. label) should be added alongside the deprecated
one.

For example, if we want to deprecate a label named `kubevirt.io/badNonIntuitiveName` to
`kubevirt.io/greatNewName`, both labels should appear.

#### Removal
Deprecated metadata will not be removed for at least 3 release cycles, which means roughly 1 year.

Afterwards, a mail should be sent to kubevirt-dev mailing list (kubevirt-dev@googlegroups.com) to inform
about the removal. If there are no objections, the metadata can be removed. Otherwise, a new removal date
can be discussed according to the context.

### API
Although the following only mentions API objects, it's also relevant for their APIs (e.g. object fields),
Kubevirt CR configuration, etc. Obviously, for fields / configs deprecation the step about adding a
deprecated annotations can be skipped. However, a warning should always be triggered for any usage of all

> of the mentioned above.

Reviewer 2: ```suggestion
of the above mentioned.
```
Maybe? Might be better to be explicit here, for clarity, even if it means repeating yourself a bit...

Author: done

#### Deprecation
In order to deprecate an API object the following actions need to be taken:
* The deprecated object needs to contain a warning about the object being deprecated. The warning should also
  state whenever the object would be removed (see below).
* A `"kubevirt.io/deprecated":"true"` annotation should be added to all of the object instances.
* A mail should be sent to kubevirt-dev mailing list (kubevirt-dev@googlegroups.com) to inform about the deprecation.

#### Removal
A deprecated object will not be removed for at least 5 release cycles, which means roughly 1.5 years.

Afterwards, a mail should be sent to kubevirt-dev mailing list (kubevirt-dev@googlegroups.com) to inform
about the removal. If there are no objections, the object can be removed. Otherwise, a new removal date
can be discussed according to the context.

> ### Feature Behavior

Reviewer 2: Maybe the API section could just say "API / Feature Behavior"? And then the removal time could say "X time (Y for Feature Behavior)". I feel it would be a bit clearer, but up to you!
Also it would be nice to have an example of some feature behavior that falls into this category, and some that don't.

Author: Hey! :)

> Maybe the API section could just say "API / Feature Behavior"? And then the removal time could say "X time (Y for Feature Behavior)". I feel it would be a bit clearer, but up to you!

The reason I prefer to separate two the is that this PR only sets the foundation for a deprecation policy. I expect that this document will grow and change as time goes. Today API and feature behavior changes are almost identical, but that can change in the future.

> Also it would be nice to have an example of some feature behavior that falls into this category, and some that don't.

Great suggestion! Added!

After the above deprecation process, the feature gate should remain as a no-op with a warning popping up.

### Metadata (e.g. labels / annotations)
Kubevirt adds certain labels / annotations to both objects that are created by us (e.g. VMs) and cluster-wide

> objects (e.g. Nodes). This metadata is exposed to the end-user which might use it and depend on it.

Reviewer 3: ```suggestion
objects (e.g. Nodes). This metadata is exposed to the end-user, which might use it and depend on it.
```

Usually, metadata needs to be deprecated if it has a non-intuitive / irrelevant name or reflects data that is no
longer relevant / useful.

#### Deprecation

> In order to deprecate metadata the following conditions must be met:

Reviewer 3: ```suggestion
In order to deprecate metadata, the following conditions must be met:
```

about the removal. If there are no objections, the metadata can be removed. Otherwise, a new removal date
can be discussed according to the context.

### API
Although the following only mentions API objects, it's also relevant for their APIs (e.g. object fields),
Kubevirt CR configuration, etc. Obviously, for fields / configs deprecation the step about adding a
deprecated annotations can be skipped. However, a warning should always be triggered for any usage of all
of the above mentioned.

#### Deprecation
In order to deprecate an API object the following actions need to be taken:
* The deprecated object needs to contain a warning about the object being deprecated. The warning should also
  state whenever the object would be removed (see below).
* A mail should be sent to kubevirt-dev mailing list (kubevirt-dev@googlegroups.com) to inform about the deprecation.

#### Removal
A deprecated object will not be removed for at least 5 release cycles, which means roughly 1.5 years.

Afterwards, a mail should be sent to kubevirt-dev mailing list (kubevirt-dev@googlegroups.com) to inform

> about the removal. If there are no objections, the object can be removed. Otherwise, a new removal date

Reviewer 3: ```suggestion
about the removal. If there are no objections, the object can be removed. Otherwise, a new target version for removal 
```

Reviewer 3: Just a thought. Perhaps it would make sense to target specific versions instead of dates?

After the above deprecation process, the feature gate should remain as a no-op with a warning popping up.

### Metadata (e.g. labels / annotations)
Kubevirt adds certain labels / annotations to both objects that are created by us (e.g. VMs) and cluster-wide
objects (e.g. Nodes). This metadata is exposed to the end-user which might use it and depend on it.

Usually, metadata needs to be deprecated if it has a non-intuitive / irrelevant name or reflects data that is no
longer relevant / useful.

#### Deprecation
In order to deprecate metadata the following conditions must be met:
* A mail should be sent to kubevirt-dev mailing list (kubevirt-dev@googlegroups.com) to inform about the deprecation.
* In case the metadata is being renamed / changed, the new metadata (e.g. label) should be added alongside the deprecated
one.

For example, if we want to deprecate a label named `kubevirt.io/badNonIntuitiveName` to
`kubevirt.io/greatNewName`, both labels should appear.

> #### Removal

Reviewer 4: We can only remove elemts from the API by increasing the major version of the PAI.

IOW- We follow semver, semver requires api compatibility within a major version of an API.
We must follow this.

Kube also picks this up in https://kubernetes.io/docs/reference/using-api/deprecation-policy/ with rule https://github.com/kubevirt/kubevirt/pull/1

Author: Thank you! A great point indeed.
Let's move the discussion to https://github.com/kubevirt/kubevirt/pull/9012#discussion_r1229514274

Reviewer 3: +1

Reviewer: > We can only remove elemts from the API by increasing the major version of the PAI.
> 
> IOW- We follow semver, semver requires api compatibility within a major version of an API. We must follow this.
> 
> Kube also picks this up in https://kubernetes.io/docs/reference/using-api/deprecation-policy/ with rule No. 1

I'll reply in the main thread above but I don't think the k8s deprecation policy suggests moving between major versions when removing elements as your suggesting here. Just that you increment the overall version. IOW dropping an element from `v1beta1` would require `v1beta2` or `v1`.

Reviewer 5: I do not think we can follow K8S rules, unless I understand them wrongly.
If we add a field to the VM spec, initially for a feature that is in alpha (not that we follow such a notion), the rule mentioned here will not allow us to revert or change anything.

I'm either not clearly grasping something basic or I do not understand how this can work.

Deprecated metadata will not be removed for at least 3 release cycles, which means roughly 1 year.

Afterwards, a mail should be sent to kubevirt-dev mailing list (kubevirt-dev@googlegroups.com) to inform
about the removal. If there are no objections, the metadata can be removed. Otherwise, a new removal date
can be discussed according to the context.

### API
Although the following only mentions API objects, it's also relevant for their APIs (e.g. object fields),
Kubevirt CR configuration, etc. Obviously, for fields / configs deprecation the step about adding a
deprecated annotations can be skipped. However, a warning should always be triggered for any usage of all
of the above mentioned.

#### Deprecation
In order to deprecate an API object the following actions need to be taken:
* The deprecated object needs to contain a warning about the object being deprecated. The warning should also
  state whenever the object would be removed (see below).

> * A `"kubevirt.io/deprecated":"true"` annotation should be added to all of the object instances.

Reviewer 6: I am slightly unsure about this. Looking at k8s, they don't seem to add this annotation to the object itself. They seem to add it to the audit record (https://kubernetes.io/docs/reference/using-api/deprecation-policy/#rest-resources-aka-api-objects).

Warning is great, about the annotation I am less sure.

Author: Yup, you're absolutely right, I've mis-read that.
Thanks for pointing that out!

* A mail should be sent to kubevirt-dev mailing list (kubevirt-dev@googlegroups.com) to inform about the deprecation.

#### Removal
A deprecated object will not be removed for at least 5 release cycles, which means roughly 1.5 years.

Afterwards, a mail should be sent to kubevirt-dev mailing list (kubevirt-dev@googlegroups.com) to inform
about the removal. If there are no objections, the object can be removed. Otherwise, a new removal date
can be discussed according to the context.

### Feature Behavior
This refers to any changes in features behavior that might surprise the user or break backwards compatibility.

As an example, let's say that a certain feature can be enabled via Kubevirt CR that affects only worker nodes,
and that feature is changed to also affect control-plane nodes. This is considered a feature behavior change.
Internal details that are not exposed to the end-user as guarantees that are part of our API aren't considered
as feature behavior changes.

The process is identical to the process of deprecating API (see above), except the number of releases before

> the final behavior change should be 3 releases (roughly 1 year).

Reviewer 7: Should we take into account the state of the Feature/API? Currently, this doc does not differentiate between alpha, beta, v1. Perhaps worth elaborating on that explicitly?

Author: Thank you! A great point indeed.
Let's move the discussion to https://github.com/kubevirt/kubevirt/pull/9012#discussion_r1229514274
